cmd/kubeadm/app/phases/upgrade/compute.go

```diff
@@ -52,38 +52,64 @@ func (u *Upgrade) CanUpgradeKubelets() bool {
 	return !sameVersionFound
 }
 
-// CanUpgradeEtcd returns whether an upgrade of etcd is possible
-func (u *Upgrade) CanUpgradeEtcd() bool {
-	return u.Before.EtcdVersion != u.After.EtcdVersion
-}
-
-// ClusterState describes the state of certain versions for a cluster
+// ClusterState describes the state of certain versions for a cluster during an upgrade
 type ClusterState struct {
 	// KubeVersion describes the version of the Kubernetes API Server, Controller Manager, Scheduler and Proxy.
 	KubeVersion string
 	// DNSVersion describes the version of the DNS add-on.
 	DNSVersion string
 	// KubeadmVersion describes the version of the kubeadm CLI
 	KubeadmVersion string
-	// KubeletVersions is a map with a version number linked to the amount of kubelets running that version in the cluster
-	KubeletVersions map[string]uint16
+	// The following maps describe the versions of the different components in the cluster.
+	// The key is the version string and the value is a list of nodes that have that version.
+	KubeAPIServerVersions         map[string][]string
+	KubeControllerManagerVersions map[string][]string
+	KubeSchedulerVersions         map[string][]string
+	EtcdVersions                  map[string][]string
+	KubeletVersions               map[string][]string
+}
+
+// AfterClusterState describes the state of certain versions for a cluster after an upgrade
+type AfterClusterState struct {
```

Comment:
can we just use the same struct ClusterState but leave the extra fields empty for "after state"?